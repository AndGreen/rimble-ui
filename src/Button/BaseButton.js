import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { buttonStyle, variant, style } from 'styled-system';
import { COMMON, LAYOUT, POSITION, BORDERS, TYPOGRAPHY } from '../constants';
import theme from '../theme';
import Icon from '../Icon';

const buttonSize = variant({
  prop: 'size',
  key: 'buttonSizes',
});

const mainColor = style({
  prop: 'mainColor',
  cssProperty: '--main-color',
  key: 'colors',
});

const contrastColor = style({
  prop: 'contrastColor',
  cssProperty: '--contrast-color',
  key: 'colors',
});

const StyledButton = styled('button')`
  & {
    -webkit-font-smoothing: antialiased;
    appearance: none;
    user-select: none;
    white-space: nowrap;
    text-decoration: none;
    text-align: center;
    overflow: hidden;
    display: inline-flex;
    align-items: center;
    justify-content: center;

    padding: ${props => (props.icononly ? '0' : props.p)};
  }

  &:hover {
    cursor: pointer;
  }

  &:disabled {
    opacity: 0.5;
    pointer-events: none;
  }

  > span.button-text {
    display: inline-flex;
  }

  ${'' /* with icon styles */}
  > div > svg,
  > span.button-text {
    position: relative;
    z-index: 1;
    user-select: none;
    pointer-events: none;
  }

  > div > svg {
    display: block;
  }
  > div:first-child > svg {
    margin-left: -0.5rem;
    margin-right: .5rem;
  }
  > div:last-child > svg {
    margin-left: .5rem;
    margin-right: -0.5rem;
  }
  > div:first-child:last-child > svg {
    margin: 0;
  }

  ${COMMON}
  ${LAYOUT}
  ${POSITION}
  ${BORDERS}
  ${TYPOGRAPHY}

  ${mainColor}
  ${contrastColor}
  ${buttonStyle}
  ${buttonSize}
`;

const ButtonBody = ({ children, icon, iconpos }) => {
  if (icon) {
    return (
      <React.Fragment>
        {icon && !iconpos && <Icon name={icon} />}
        {icon && iconpos == 'left' && <Icon name={icon} />}
        {children && <span className="button-text" children={children} />}
        {icon && iconpos == 'right' && <Icon name={icon} />}
      </React.Fragment>
    );
  } else {
    return <span className="button-text" children={children} />;
  }
};

const Button = React.forwardRef((props, ref) => {
  return <StyledButton {...props} ref={ref} />;
});

StyledButton.defaultProps = {
  theme,
};

Button.defaultProps = {
  // base props
  // variant: 'primary',
  // size: 'medium',
  position: 'relative',
  height: '3rem',
  minWidth: '3rem',
  m: 0,
  px: 4,
  py: 0,
  fontSize: 'inherit',
  fontFamily: 'sansSerif',
  fontWeight: 3,
  lineHeight: 1,
  // color props
  color: 'white',
  bg: 'silver',
  mainColor: 'primary',
  contrastColor: 'white',
  border: 'none',
};

Button.propTypes = {
  ...COMMON.propTypes,
  ...LAYOUT.propTypes,
  ...POSITION.propTypes,
  ...BORDERS.propTypes,
  ...TYPOGRAPHY.propTypes,
  /**
   * Sets theme
   */
  theme: PropTypes.object,
  /**
   * Sets background color of button
   */
  mainColor: PropTypes.string,
  /**
   * Sets text color of button
   */
  contrastColor: PropTypes.string,
  /**
   * Sets font and spacing size of button
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * Sets icon of button
   */
  icon: PropTypes.string,
  /**
   * Sets icon position on button
   */
  iconpos: PropTypes.oneOf(['left', 'right']),
  /**
   * Sets padding on button with icon and no text
   */
  icononly: PropTypes.bool,
};

Button.displayName = 'Button';

export { StyledButton, ButtonBody };

export default Button;
