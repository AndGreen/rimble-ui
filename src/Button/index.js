import React from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'

import { tint, shade } from 'polished'

import {
  space,
  color,
  bgColor,
  fontWeight ,
  borders,
  borderColor,
  borderRadius,
  boxShadow
} from 'styled-system'

import theme from '../theme'

const fullWidth = (props) => (
  props.fullWidth ? { width: '100%' } : null
)

const size = (props) => {
  switch (props.size) {
    case 'small':
      return `
        font-size: 12px;
        height: 24px;
      `
    case 'medium':
      return `
        font-size: 16px;
        height: 48px;
      `
    case 'large':
      return `
        font-size: 24px;
        height: 64px;
      `
    default:
      return `
        font-size: 16px;
        height: 48px;
      `
  }
}

const Button = styled.button`
  -webkit-font-smoothing: antialiased;
  position: relative;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  text-decoration: none;
  font: inherit;
  line-height: 1;
  color: inherit;
  cursor: pointer;
  white-space: nowrap;
  border-width: 0;
  border-style: solid;

  box-shadow:
    0 4px 6px rgba(50,50,93,.11),
    0 1px 3px rgba(0,0,0,.08)
  ;
  transition: all .15s ease;
  transform-origin: center;

  ${size} ${fullWidth}

  ${space}
  ${color}
  ${bgColor}
  ${fontWeight}
  ${borders}
  ${borderColor}
  ${borderRadius}

  &:hover {
    background-color: ${props => tint(0.1, props.theme.colors.primary)};
    transform: translateY(-1px);
    box-shadow:
      0 7px 14px rgba(50,50,93,.1),
      0 3px 6px rgba(0,0,0,.08)
    ;
  }
  &:disabled {
    opacity: 0.25;
    pointer-events: none;
    cursor: not-allowed;
  }
  &:active {
    background-color: ${props => shade(0.1, props.theme.colors.primary)};
    box-shadow: none;
  }
`

Button.propTypes = {
  ...fontWeight.propTypes,
  ...borders.propTypes,
  ...borderColor.propTypes,
  ...borderRadius.propTypes,
  // ...buttonStyle.propTypes,
}

Button.defaultProps = {
  as: 'button',
  fontSize: 'inherit',
  fontWeight: 'bold',
  m: 0,
  px: 3,
  py: 0,
  color: 'white',
  bg: 'primary',
  border: 'none',
  borderRadius: 0,
  theme: theme
}

Button.displayName = 'Button'

export default Button
