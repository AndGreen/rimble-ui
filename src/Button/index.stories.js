import { storiesOf } from '@storybook/react'

import React from 'react'
import Button from '../Button'

storiesOf('Button', module)
  .add('sizes', () => (
    <div>
      <Button size='small'>Small Button</Button>
      <Button size='medium'>Medium Button</Button>
      <Button size='large'>Large Button</Button>
    </div>
  ))
  .add('full width', () => (
    <div>
      <Button fullWidth>Full-width Button</Button>
    </div>
  ))
  .add('modes', () => (
    <div>
      <Button mode='solid'>Solid Button</Button>
      <Button mode='outline'>Outline Button</Button>
      <Button mode='text'>Text Button</Button>
    </div>
  ))
