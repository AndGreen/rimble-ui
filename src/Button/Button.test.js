import Button from './'

describe('Button component sanity', () => {
  it('has name', () => {
    expect(Button.displayName).toBe('Button')
  })

  it('matches default snapshot', () => {
    const component = render(<Button />)
    expect(component).toMatchSnapshot()
  })
})
