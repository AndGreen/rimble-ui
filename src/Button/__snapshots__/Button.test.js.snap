// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component sanity matches default snapshot 1`] = `
.c1 {
  margin: 0px;
  padding-left: 32px;
  padding-right: 32px;
  padding-top: 0px;
  padding-bottom: 0px;
  color: #fff;
  background-color: #4E3FCE;
  font-size: inherit;
  height: 48px;
  border: 0;
  border-color: none;
  border-radius: 4px;
}

.c1 {
  box-sizing: border-box;
}

.c0 {
  font-weight: 600;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.1);
  font-size: 1rem;
  min-width: 48px;
}

.c0 {
  font-family: inherit;
  position: relative;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  line-height: 1;
  overflow: hidden;
  white-space: nowrap;
  -webkit-font-smoothing: antialiased;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all .15s ease;
  transition: all .15s ease;
}

.c0:hover {
  background-color: #5f52d2;
  box-shadow: 0px 2px 4px rgba(0,0,0,0.2);
}

.c0:active {
  background-color: #4638b9;
  box-shadow: none;
}

.c0:disabled {
  opacity: 0.4;
  cursor: not-allowed;
}

.c0 > svg {
  display: block;
}

.c0 > svg:first-child {
  margin-left: -1rem;
  margin-right: .5rem;
}

.c0 > svg:last-child {
  margin-left: .5rem;
  margin-right: -1rem;
}

.c0 > svg:first-child:last-child {
  margin: 0;
}

<button
  class="c0 c1"
  color="white"
  font-size="inherit"
  font-weight="3"
  height="48px"
  type="button"
>
  
</button>
`;
